drnn.py

```diff
@@ -1,91 +1,305 @@
+                            'test_acc': [], 'val_loss': [], 'val_acc': []}
+
+            # building the rest of the model
+            self.logits
+            self.optimize
+
+            # initializing the model's instance to be used and restoring the
+            # weights from a file if passed with a value
+            gpu_options = tf.GPUOptions(per_process_gpu_memory_fraction=gpu_fraction)
+            session_conf = tf.ConfigProto(
+                allow_soft_placement=True,
+                log_device_placement=False,
+                gpu_options=gpu_options)
+            self.current_sess = tf.Session(config=session_conf)
+
+            init_g = tf.global_variables_initializer()
+            init_l = tf.local_variables_initializer()
+            if restore_file:
+                saver = tf.train.Saver()
+                saver.restore(self.current_sess, restore_file)
+            else:
+                self.current_sess.run(init_g)
+                self.current_sess.run(init_l)
+
+    @define_scope
+    def logits(self):
         # embedding layer
         with tf.device('/cpu:0'), tf.name_scope('embedding'):
-            self.W = tf.Variable(tf.random_uniform([vocab_size, embedding_size], -1.0, 1.0, name="W"))
-            self.embedding_placeholder = tf.placeholder(tf.float32, [vocab_size, embedding_size])
-            self.embedding_init = self.W.assign(self.embedding_placeholder)
-            self.embedding_vectors = tf.nn.embedding_lookup(self.W, self.input_x)
-            # print(self.embedding_vectors.shape)
+            word_embeddings = tf.get_variable(name="Embeddings_mat_word",
+                                              shape=self.word_embeddings_mat.shape,
+                                              initializer=tf.constant_initializer(
+                                                  self.word_embeddings_mat),
+                                              trainable=True)
+
+            embedding_vectors = tf.nn.embedding_lookup(word_embeddings, self.word_ids)
 
         with tf.name_scope('dropout'):
-            self.embedding_vectors_dropout = tf.nn.dropout(self.embedding_vectors, self.dropout_keep_prob)
+            embedding_vectors_dropout = tf.nn.dropout(embedding_vectors, self.dropout_keep_prob)
 
         with tf.name_scope("DGRU_MLP"):
-            self.hidden = []
-            self.input = tf.pad(self.embedding_vectors_dropout, [[0, 0], [num_k-1, 0], [0, 0]])
-            # print(self.input.shape)
+            hidden = []
+            logging.info('embeddings dropout: {}'.format(embedding_vectors_dropout.shape))
+            input = tf.pad(embedding_vectors_dropout, [[0, 0], [self.num_k-1, 0], [0, 0]])
+            logging.info('embeddings padded: {}'.format(input.shape))
+
             start = 0
-            end = start + num_k - 1
-            while end < (sequence_length + num_k - 1):
-                input_k = self.input[:, start:end, :]
+            end = start + self.num_k - 1
```

Comment:
I think this is not correct. According to section 3.2 in the paper, the RNN input should be the current token preceded with the previous _k - 1_ tokens. with the way `end` is defined here, it takes only the previous _k - 1_ tokens and not the current token itself.

Imagine we have the sentence (I have a cat) and a _k = 3_, we're going to pad this sentence with two 0 tokens at the beginning according to the `tf.pad` statement a few lines above, this will result in the sentence `0 0 I have a cat` whose tokens have the indices `0 1 2 3 4 5`. Now when we want to get the hidden state of the token `have`, `start` will equal `1` and `end` (according to the original implementation) will be `3`. The slice `start:end` will choose all the tokens beginning at `start` and ending at `end - 1` (this is python's convention and TF's convention [AFAIK](https://www.tensorflow.org/versions/r1.14/api_docs/python/tf/Tensor#__getitem__)) which will select tokens `1 2` which translates to `0 I`, not including `have`. The correct way to do it is: 
```suggestion
            end = start + self.num_k
```

Reply:
Good catch, fixed.